clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1062,8 +1062,13 @@ private void close(long timeout, TimeUnit timeUnit, boolean swallowException) {
         ClientUtils.closeQuietly(partitioner, "producer partitioner", firstException);
         AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId);
         log.debug("Kafka producer has been closed");
-        if (firstException.get() != null && !swallowException)
-            throw new KafkaException("Failed to close kafka producer", firstException.get());
+        Throwable exception = firstException.get();
+        if (exception != null && !swallowException) {
```

Comment:
Wondering if we should raise the the interrupt exception regardless of `swallowException`?

Reply:
Hmmm... `swallowException` is only set if `close()` is called from constructor. Thus, is we don't swallow, we overwrite the original root cause exception that did make the constructor fail as the second throw replaces the first one. So I guess it's ok as is?

Reply:
Yes, I was thinking about that. If the intent is to make sure that interrupts are seen by the caller, it seems like losing the configuration exception may be the price we have to pay (though we could at least log it). Probably not too big of a deal since we're unlikely to need to join the io thread if there is some error during construction.

Reply:
So what are you suggesting? Keep as is? Or rethrow even if swallow is set?

Reply:
@hachikuji Ping

Reply:
@mjsax Maybe if `swallowException` is set, maybe we should just reset the interrupt with `Thread.currentThread().interrupt()`? At least that way we aren't covering up the fact that an interrupt happened.

By the way, it would be nice to have a test case at least for the common case

Reply:
Ack. Will add this for the Consumer, too, than.

Reply:
Had a look into testing. It seems that `InterruptedException` for consumer can only occur within the coordinator when it does an commit on close. But `swallowException` can only be set to `true` within the constructor, thus, it can never happen to get an `InterruptedException` if `swallowException=true`.

For the Producer, it could only happen on senderThread.join -- the sender thread only blocks within the network client in `this.nioSelector.select(ms);` -- but not sure if this can be interrupted in practice? And if yes, I am not sure, how to write a test for this? I guess, we would need to add a new package private constructor to `KafakProducer` similar to `KafkaConsumer` to be able to pass in some mocks... Not sure if it's worth it...

Reply:
I think it's probably fine if we don't cover the `swallowException` cases. A normal `close()` is more likely in practice, so it would be good if we could cover that case.

Reply:
Consumer is covered already: https://github.com/apache/kafka/blob/trunk/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java#L1237

Can add a similar test for Producer to inject an interrupt there.